_posts/2020-09-16-hibernate-search.md

```diff
@@ -0,0 +1,440 @@
+We can still perform sorting operations over other fields which are not marked with this annotation but that will have some
+performance penalties.
+
+6. `@IndexedEmbedded` is used when we want to perform a search over nested objects fields. For instance, let's say we want to search
+all posts made by a user with the first name 'Joe' (user.first: joe). 
+
+
+Some things to note about analyzers and normalizers when using Elasticsearch integration:
+* Although you can use Lucene's built-in analyzers or even custom analyzers using `@AnalyzerDef`, it's recommended to use 
+built-in Elasticsearch analyzers.
+* Built-in Normalizers are directly translated to Elasticsearch supported normalizers but for this to work you 
+need to define Normalizer using `@NormalizerDef` and pass the name in definition as I have done in `Tag` enum's field.
+
+We need to cover a few more properties of `@Field` annotation before moving on. Let's look at `User` class for that.
+
+```java
+@Getter
+@Setter
+@Entity
+@Indexed(index = "idx_user")
+class User {
+  @Id
+  private String id;
+
+  @Field(store = Store.YES)//->[3]
+  @Field(name = "fullName")//->[2]
+  private String first;
+
+  @Field(index = Index.NO) //->[1]
+  private String middle;
+
+  @Field
+  @Field(name = "fullName")//->[2]
+  private String last;
+
+  @Field
+  private Integer age;
+
+  @ContainedIn//->[4]
+  @OneToMany(mappedBy = "user")
+  private List<Post> post;
+}
+```
+
+Couple of interesting things to notice here. Let's start with the `index` property.
+
+1. `Index.NO` this indicates that property will be stored in the index document, but it won't be indexed. We won't be able
+to perform any search operation over it. You might be thinking "Why not simply remove `@Field` annotation?". And the answer is that we still need this field for Projection (Covered in the next section).
+
+2. In the `Post` class we saw that we can map one entity field to multiple index document fields. We can also do the inverse. As 
+you can see `@Field(name = "fullName")` is mapped to `first` and `last` both. This way `fullName` will have content of both fields. So, instead of searching over the `first` and `last` field separately, we can directly search over `fullName`. 
+
+3. We can set store to `Store.YES` when we plan to use it in projection. Note that this will require extra space. Plus, 
+Elasticsearch already stores the value in the `_soucre` field. So, the only reason to set the store to true is that when we don't want Elasticsearch to look up and extract value from the `_source` field.
+
+4. `@ContainedIn` just like `@OneToMany`-`@ManyToOne` annotation makes relationship bidirectional. When the values of this
+entity will be updated its values in the root entity i.e., in the `idx_post`(Post) index will also be updated.
+
+## Performing Queries
+Before we perform any queries, we first need to load data into the Elasticsearch.
```

Comment:
A sentence like "Before we do X" is a clear sign that this should be its own subsection.